### ISSUE TITLE
Whether corner_pr is softmax normalized probability or log of it?

### ISSUE THREAD

**Issue by ghost** [lachlants/denet#5]
Hi @lachlants, thanks for your paper and code. I really enjoy using it.

I have one issue regarding the corner_pr in denet_sparse.py. It seems to me [here](https://github.com/lachlants/denet/blob/master/denet/layer/denet_sparse.py#L141) the corner_pr is the probability map after softmax normalization. And self.corner_threshold should be the probability threshold. However, in your C implementation, the probability is summed instead of multiplied (https://github.com/lachlants/denet/blob/master/denet/layer/denet_sparse.cc#L271) and you also took a log on the threshold https://github.com/lachlants/denet/blob/master/denet/layer/denet_sparse.cc#L506. 

I believe I miss something. Could you help me point out it?

**Comment by lachlants** [comment]
Hello, good to see people get some use out of the codebase. 

The **corner_pr** variable is actually the **log** probability that a corner exists at (x,y)  i.e. log(Pr(c|x,y)). Using the log probability allows us to turn multiplication into addition via the identity log(AB) = log(A) + log(B), therefore as a small optimization I keep all the probabilities in the logarithmic basis (it's also needed for the cross entropy training loss). In particular I make the following substitution:

![codecogseqn](https://user-images.githubusercontent.com/7444608/34132402-c48b35d6-e4a3-11e7-81bb-e94295bb7f79.gif)

**Comment by ghost** [comment]
Hi @lachlants , thanks for your reply. I can understand it now.

I'm currently confused about this [code block](https://github.com/lachlants/denet/blob/master/denet/layer/denet_corner.py#L50-L53). 
It seems to me that the softmax is calculated over self.corner_lh and -self.corner_lh. And self.corner_lh is the last 4 (or 5) channels of a convolution's output. Do I understand correctly in this point?

Another question is about the denet_sparse.cc again. I didn't understand this line: https://github.com/lachlants/denet/blob/master/denet/layer/denet_sparse.cc#L306. Is the pr_f the log of background probability or foreground probability? 
I guess you want to make something like: 1 / (1 + e^(logp - log(1-p))). (say p is the foreground probability). But I'm confused. First, the equation will output 1-p instead of p. Second, 1 / (1 + e^(log(p1*p2*p3*p4) - log( 1-p1 * 1-p2 * 1-p3 * 1-p4) )) seems meaningless. 

Again, I believe I missed something. Could you help me point out it?

Thanks a lot for your help!

**Comment by lachlants** [comment]
> It seems to me that the softmax is calculated over self.corner_lh and -self.corner_lh. And self.corner_lh is the last 4 (or 5) channels of a convolution's output. Do I understand correctly in this point?

Yep, thats correct. The unnormalized log likelihood of a corner existing is given by **-x**, whereas the log likelihood of a corner not existing is given by **+x** , where **x** is the output of the convolutional layer. Using the positive and negative values for true / false means the network only has to output a single value and is mathematically equivalent as having a value for each state. 

> Another question is about the denet_sparse.cc again. I didn't understand this line: https://github.com/lachlants/denet/blob/master/denet/layer/denet_sparse.cc#L306. Is the pr_f the log of background probability or foreground probability?

pr_f = background (false) log likelihood, pr_t = foreground (true) log likelihood

> I guess you want to make something like: 1 / (1 + e^(logp - log(1-p))). (say p is the foreground probability). But I'm confused. First, the equation will output 1-p instead of p. Second, 1 / (1 + e^(log(p1p2p3*p4) - log( 1-p1 * 1-p2 * 1-p3 * 1-p4) )) seems meaningless.

This equation is accounting for the fact that the probability is no longer normalized. e.g. the equation I gave above should really be:

![codecogseqn 3](https://user-images.githubusercontent.com/7444608/34138551-3a9d8d6c-e4c4-11e7-8fcd-e85236657870.gif)

where L_1 is the log likelihood of a foreground object, L_0 is the log likelihood of a background object, Pr(s|c,y,x) is the probability of a corner of type c at (x,y) and Pr(bbox) is the normalized probability of a foreground object. I suspect the abs() is an error... however its very unlikely to affect anything since pr_f >> pr_t in most cases.

**Comment by ghost** [comment]
Thanks a lot!

For the first part:
I see, so if I understand correctly, [this code](https://github.com/lachlants/denet/blob/master/denet/layer/denet_corner.py#L42-L47) will initialize the convolution (or BatchNorm?) with a positive bias and zero weight. I guess this may because of imbalance number of bg/fg?

For the second part:
So the probability value of a box is actually not the product of 4 corner probabilities. But you map the value by a monotonic function and thus should be the same during sorting?

**Comment by lachlants** [comment]
> I see, so if I understand correctly, this code will initialize the convolution (or BatchNorm?) with a positive bias and zero weight. I guess this may because of imbalance number of bg/fg?

Yep correct, it gets much faster convergence with this initialization. 

> So the probability value of a box is actually not the product of 4 corner probabilities. But you map the value by a monotonic function and thus should be the same during sorting? 

The probability of the box is **proportional** to (but not equal to) the product of the 4 corner probabilities. You have to apply the normalization mentioned above to recover the probability.

**Comment by ghost** [comment]
Thanks! I have a deeper understanding now.

I closed the issue since my problem is solved.

**Comment by ghost** [comment]
Hi, I re-thought the 'error' you mentioned in

> I suspect the abs() is an error... however its very unlikely to affect anything since pr_f >> pr_t in most cases.

I'm a little bit confused about why you say this is not to affect anything. In my opinion, the positive RoIs will have pr_t > pr_f and thus the probability is incorrect. 

Do you have any updated experiments for this issue?

**Comment by lachlants** [comment]
> I'm a little bit confused about why you say this is not to affect anything. In my opinion, the positive RoIs will have pr_t > pr_f and thus the probability is incorrect.

> Do you have any updated experiments for this issue?

Yep I would agree, however in practice it doesn't appear to changes things much (at least in terms of MAP@IoU=0.5). I've tested it by fixing the bug and performed inference on Pascal VOC. It's possible it might help in other regards (e.g. training, reducing the number of RoIs needed, MAP @ higher IoUs, etc) but I haven't tested them.
